File: lingting-core/src/main/java/live/lingting/framework/value/cycle/IteratorCycleValue.java
package live.lingting.framework.value.cycle;

import live.lingting.framework.value.CycleValue;
import live.lingting.framework.value.StepValue;

import java.util.Iterator;

/**
 * @author lingting 2024-01-23 15:24
 */
public class IteratorCycleValue<T> extends CycleValue<T> {

	public IteratorCycleValue(Iterator<T> iterator) {
		super(new StepCycleValueFunction<>(StepValue.iterator(iterator)));
	}

}

File: lingting-core/src/main/kotlin/live/lingting/framework/value/cycle/IteratorCycleValue.kt
package live.lingting.framework.value.cycle

import live.lingting.framework.value.step.IteratorStepValue

/**
 * @author lingting 2024-01-23 15:24
 */
class IteratorCycleValue<T>(private val step: IteratorStepValue<T>) : AbstractConcurrentCycleValue<T>() {

    constructor(iterable: Iterable<T>) : this(iterable.iterator())

    constructor(iterator: Iterator<T>) : this(IteratorStepValue<T>(iterator))

    override fun doReset() {
        step.reset()
    }

    override fun doNext(): T {
        if (!step.hasNext()) {
            step.reset()
        }
        return step.next()
    }

    /**
     * 移除上一个next返回的元素
     */
    fun remove() {
        step.remove()
    }

}
